### PR TITLE
[GCI-prathamesh_m009] Made connection to Database

### PR DESCRIPTION
**Pull request by prathamesh-mutkure**
This PR is part of GoLang-2 Task

**Task Instance:** https://codein.withgoogle.com/dashboard/task-instances/5961818246217728/

## /patient/all
Patient-1
Patient-2
Patient-3

## /location/all
Patient-1

## /hospital/all
Patient-1

### CODE REVIEW

**Comment by prathamesh-mutkure** [comment]
- Moved all the API Handlers to seperate `api.go` file
- Added screenshots in `/docs/` directory and also in PR
- For adding patients in database, I copied names from our GCI Telegram chat for convenience

Thank You :)

File: scripts/db_schema.sql [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+INSERT INTO patient (name, illness, birth_date, location_id)
+VALUES ('Ray Ma', 'Fever', '2000-05-15', 1),
+       ('Tejas Mate', 'Cold', '2004-10-12', 2),
+       ('Ribhav Sharma', 'Diabetes', '1990-08-01', 3),
+       ('Hrishikesh Patil', 'Asthama', '1981-12-23', 4),
+       ('Nayan Jindal', 'Flu', '2006-01-30', 1),
+       ('Ishaan Das', 'Dengue', '2003-11-19', 2),
+       ('Mihir Aggarwal', 'Cough', '1995-03-07', 3),
+       ('Khushi Walia', 'Cold', '1985-06-01', 4),
+       ('Inesh Tickoo', 'Malaria', '2009-10-31', 1);
```
Comment (PrathameshMutkure) [inline scripts/db_schema.sql]: If these names seem familier to you, sorry, I copied them from our GCI Telegram chat for convinience :)

File: api.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	name := c.Param("name")
+	c.JSON(http.StatusOK, gin.H{"greet-message": "Hello " + name + "!"})
+}
+
+/*
+ * Returns a json array of all the products
+ */
+func GetAllPatients(c *gin.Context) {
+
+	rows, err := db.Query("SELECT * FROM patient")
```
Comment (f4ww4z) [inline api.go]: You should be using `sqlx` for this. The `db` should be of type `*sqlx.DB`, so you can do `db.Select()`, which is easier to read. https://jmoiron.github.io/sqlx/

File: api.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+		}
+		patients = append(patients, row)
+	}
+
+	err = rows.Err()
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	c.JSON(http.StatusOK, gin.H{"patients": patients})
```
Comment (f4ww4z) [inline api.go]: No need to wrap with `"patients"` attribute, just return the `patients` array.

File: api.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+/*
+ * Returns a json array of all the locations
+ */
+func GetAllLocations(c *gin.Context) {
+	rows, err := db.Query("SELECT * FROM location")
```
Comment (f4ww4z) [inline api.go]: See above comment. Should be using `sqlx.DB#Select()`

File: api.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+		}
+		locations = append(locations, row)
+	}
+
+	err = rows.Err()
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	c.JSON(http.StatusOK, gin.H{"locations": locations})
```
Comment (f4ww4z) [inline api.go]: No need to wrap with `"locations"` attribute, just return the `locations` array.

File: api.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+/*
+ * Returns json array of all the hospitals
+ */
+func GetAllHospitals(c *gin.Context) {
+	rows, err := db.Query("SELECT * FROM hospital")
```
Comment (f4ww4z) [inline api.go]: See above comment.

File: api.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+		}
+		hospitals = append(hospitals, row)
+	}
+
+	err = rows.Err()
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	c.JSON(http.StatusOK, gin.H{"hospitals": hospitals})
```
Comment (f4ww4z) [inline api.go]: No need to wrap with `"hospitals"` attribute, just return the `hospitals` array.

File: main.go [hunk 8]
```diff
@@ -1,31 +1,76 @@
...
+
+const (
+	DB_USER     = "twvjbojy"
+	DB_PASSWORD = "H0kJZUiDgpgweAWNWdLRgI6v6-IEyxyr"
+	DB_NAME     = "twvjbojy host=john.db.elephantsql.com"
+	DB_HOST     = "john.db.elephantsql.com"
 )
 
 var router *gin.Engine
+var db *sql.DB
```
Comment (f4ww4z) [inline main.go]:
```suggestion
var db *sqlx.DB
```

File: main.go [hunk 9]
```diff
@@ -1,31 +1,76 @@
 package main
 
 import (
+	"database/sql"
```
Comment (f4ww4z) [inline main.go]: Shouldn't need this package.

File: main.go [hunk 10]
```diff
@@ -1,31 +1,76 @@
...
 )
 
 var router *gin.Engine
+var db *sql.DB
 
-func main() {
-	router = gin.Default()
-	initializeRoutes()
-	router.Run()	// Runs on default port 8080
+type patient struct {
```
Comment (f4ww4z) [inline main.go]: Export this struct by capitalizing the first letter. 
```suggestion
type Patient struct {
```

File: main.go [hunk 11]
```diff
@@ -1,31 +1,76 @@
...
+	ID         int    `json:"id"`
```
Comment (f4ww4z) [inline main.go]: Can you add the database attribute name too? Do this for each attribute. E.g.:
```suggestion
	ID         int    `json:"id" db:"id"`
```

File: main.go [hunk 12]
```diff
@@ -1,31 +1,76 @@
...
+	Illness    string `json:"illness"`
+	BirthDate  string `json:"birth_date"`
+	LocationId int    `json:"location_id"`
 }
 
-// This function will initialize all the routes
-func initializeRoutes() {
-	router.GET("/hello", helloHandler)
-	router.GET("/greet/:name", greetHandler)
+type location struct {
```
Comment (f4ww4z) [inline main.go]: Export this struct by capitalizing the first letter.

File: main.go [hunk 13]
```diff
@@ -1,31 +1,76 @@
...
+	ID         int    `json:"id"`
```
Comment (f4ww4z) [inline main.go]: Add the database representation / name for all the attributes.

File: main.go [hunk 14]
```diff
@@ -1,31 +1,76 @@
...
+	HospitalId int    `json:"hospital_id"`
 }
 
-// Returns a HelloWorld json message
-func helloHandler(c *gin.Context) {
-	c.JSON(http.StatusOK, gin.H{ "message": "Hello World!"})
+type hospital struct {
+	ID         int    `json:"id"`
+	Name       string `json:"name"`
+	HospitalId int    `json:"max_patient_amount"`
```
Comment (f4ww4z) [inline main.go]: This attribute name doesn't seem to match. Also, add the database representation / name for all the attributes. 
```suggestion
	MaxPatientAmount int    `json:"max_patient_amount" db:"max_patient_amount"`
```

**Review by f4ww4z — CHANGES_REQUESTED** [review]
Good start @PrathameshMutkure . See my comments.

File: api.go (hunk 5 above)
Comment (prathamesh-mutkure) [reply api.go]: Done

File: main.go (hunk 10 above)
Comment (prathamesh-mutkure) [reply main.go]: Sorry, not aware of best practices in Go, will try to learn them on the go :)

File: main.go (hunk 14 above)
Comment (prathamesh-mutkure) [reply main.go]: Sorry, corrected it :)

File: main.go (hunk 11 above)
Comment (prathamesh-mutkure) [reply main.go]: Thanks for this one :)

File: main.go (hunk 13 above)
Comment (prathamesh-mutkure) [reply main.go]: Done

File: main.go (hunk 9 above)
Comment (prathamesh-mutkure) [reply main.go]: Done

File: api.go (hunk 2 above)
Comment (prathamesh-mutkure) [reply api.go]: Thanks for it, it reduced my coding errorts and lines to almost half :)

File: api.go (hunk 4 above)
Comment (prathamesh-mutkure) [reply api.go]: It made my work way more easy, Thanks :)

**Review by f4ww4z — APPROVED** [review]

**Pull request merged**